src/ai/ai_gui.cpp

```diff
@@ -264,6 +267,354 @@ struct AIConfigWindow : public Window {
+	{
+		if (this->timeout.Elapsed(delta_ms)) {
+			this->clicked_button = -1;
+			this->SetDirty();
+		}
+	}
+
+	/**
+	 * Some data on this window has become invalid.
+	 * @param data Information about the changed data.
+	 * @param gui_scope Whether the call is done from GUI scope. You may not do everything when not in GUI scope. See #InvalidateWindowData() for details.
+	 */
+	void OnInvalidateData(int data = 0, bool gui_scope = true) override
+	{
+		this->RebuildVisibleSettings();
+		HideDropDownMenu(this);
+		this->CloseChildWindows(WC_QUERY_STRING);
+	}
+
+private:
+	bool IsEditableItem(const ScriptConfigItem &config_item) const
+	{
+		return _game_mode == GM_MENU
+			|| _game_mode == GM_EDITOR
+			|| !Company::IsValidID(this->slot)
+			|| (config_item.flags & SCRIPTCONFIG_INGAME) != 0
+			|| _settings_client.gui.ai_developer_tools;
+	}
+
+	void SetValue(int value)
+	{
+		VisibleSettingsList::const_iterator it = this->visible_settings.begin();
+		for (int i = 0; i < this->clicked_row; i++) it++;
+		const ScriptConfigItem config_item = **it;
+		if (_game_mode == GM_NORMAL && Company::IsValidID(this->slot) && (config_item.flags & SCRIPTCONFIG_INGAME) == 0) return;
+		this->script_config->SetSetting(config_item.name, value);
+		this->SetDirty();
+	}
+};
+
+/** Widgets for the Script settings window. */
+static const NWidgetPart _nested_ai_settings_widgets[] = {
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_CLOSEBOX, COLOUR_MAUVE),
+		NWidget(WWT_CAPTION, COLOUR_MAUVE, WID_AIS_CAPTION), SetDataTip(STR_AI_SETTINGS_CAPTION_AI, STR_TOOLTIP_WINDOW_TITLE_DRAG_THIS),
+		NWidget(WWT_DEFSIZEBOX, COLOUR_MAUVE),
+	EndContainer(),
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_MATRIX, COLOUR_MAUVE, WID_AIS_BACKGROUND), SetMinimalSize(188, 182), SetResize(1, 1), SetFill(1, 0), SetMatrixDataTip(1, 0, STR_NULL), SetScrollbar(WID_AIS_SCROLLBAR),
+		NWidget(NWID_VSCROLLBAR, COLOUR_MAUVE, WID_AIS_SCROLLBAR),
+	EndContainer(),
+	NWidget(NWID_HORIZONTAL),
+		NWidget(NWID_HORIZONTAL, NC_EQUALSIZE),
+			NWidget(WWT_PUSHTXTBTN, COLOUR_MAUVE, WID_AIS_ACCEPT), SetResize(1, 0), SetFill(1, 0), SetDataTip(STR_AI_SETTINGS_CLOSE, STR_NULL),
+			NWidget(WWT_PUSHTXTBTN, COLOUR_MAUVE, WID_AIS_RESET), SetResize(1, 0), SetFill(1, 0), SetDataTip(STR_AI_SETTINGS_RESET, STR_NULL),
+		EndContainer(),
+		NWidget(WWT_RESIZEBOX, COLOUR_MAUVE),
+	EndContainer(),
+};
+
+/** Window definition for the AI settings window. */
+static WindowDesc _ai_settings_desc(
+	WDP_CENTER, "settings_ai", 500, 208,
+	WC_AI_SETTINGS, WC_NONE,
+	0,
+	_nested_ai_settings_widgets, lengthof(_nested_ai_settings_widgets)
+);
+
+/**
+ * Open the AI settings window to change the settings for a Script.
+ * @param slot The CompanyID of the Script to change the settings.
+ */
+void ShowAISettingsWindow(CompanyID slot)
+{
+	CloseWindowByClass(WC_SCRIPT_LIST);
+	CloseWindowByClass(WC_AI_SETTINGS);
+	new AISettingsWindow(&_ai_settings_desc, slot);
+}
+
 /** Open the AI config window. */
```

Comment:
I think we usually try to keep the `ShowXXXWindow()` near the window definition.